Ensure mirror pods are created as soon as node is registered

AnishShah opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Mirror pods for static pods may not be created immediately during node startup 
because either the node is not registered or node informer is not synced.
They will be created eventually  when static pods are resynced (every 1-1.5 minutes).

However, during this delay of 1-1.5 mins, kube-scheduler might overcommit resources
to the node and eventually cause kubelet to reject pods with
 OutOfCPU/OutOfMemory/OutOfPods error.

To ensure kube-scheduler is aware of static pod resource usage faster, 
mirror pods are created as soon as the node registers.

#### Which issue(s) this PR fixes:

ref #115325

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

bart0sh commented on pkg/kubelet/kubelet.go:
```diff
@@ -2244,6 +2244,27 @@ func (kl *Kubelet) getPodsToSync() []*v1.Pod {
+			return
+		}
+		close(nodeCh)
+	}, time.Second, nodeCh)
```
@AnishShah This will block Kubelet until GetNode succeeds. Is it ok from your point of view?

bart0sh commented on pkg/kubelet/kubelet.go:
```diff
@@ -2244,6 +2244,27 @@ func (kl *Kubelet) getPodsToSync() []*v1.Pod {
+		close(nodeCh)
+	}, time.Second, nodeCh)
+
+	allPods := kl.podManager.GetPods()
+	staticPods := []*v1.Pod{}
+	for _, pod := range allPods {
+		if kubetypes.IsStaticPod(pod) {
+			staticPods = append(staticPods, pod)
+		}
+	}
```
nit:
```suggestion
        _, staticPods := splitPodsByStatic(kl.podManager.GetPods())
```

bart0sh replied:
alternatively you can use [slices.DeleteFunc](https://pkg.go.dev/slices#DeleteFunc), but that would be more wordy.

bart0sh commented:
/triage accepted

As this is a bugfix, please provide unit or e2e test(s) covering it.

liggitt commented:
Thanks for looking into this. There's some tricky/fragile ordering around kubelet startup / registration / pod source syncing we need to ensure this correctly integrates with.

/assign @smarterclayton @bobbypage 
to make sure this aligns with kubelet pod lifecycle

liggitt commented:
some general questions:
* should completion of initial registration of mirror pods be connected to self-registration on first start, or to the node marking itself as ready?
* should initial registration of mirror pods be best-effort and just continue to mark ready on failure (which I *think* the current PR does), or block node readiness?
* there's a *lot* of stuff that gets started in Kubelet#Run before kicking off the normal `kl.syncLoop`, which previously was the only thing that called HandlePodSyncs ... do we know that it is safe to add a new call site to HandlePodSyncs from outside the sync loop before the stuff in Kubelet#Run has happened?

Rather than moving static pod registration, I sort of expected us to just track whether the existing code paths had registered static pods at least once, and gate reporting the node as ready until that happened here:

```go
// ReadyCondition returns a Setter that updates the v1.NodeReady condition on the node.
func ReadyCondition(
	nowFunc func() time.Time, // typically Kubelet.clock.Now
	runtimeErrorsFunc func() error, // typically Kubelet.runtimeState.runtimeErrors
	networkErrorsFunc func() error, // typically Kubelet.runtimeState.networkErrors
	storageErrorsFunc func() error, // typically Kubelet.runtimeState.storageErrors
	cmStatusFunc func() cm.Status, // typically Kubelet.containerManager.Status
	nodeShutdownManagerErrorsFunc func() error, // typically kubelet.shutdownManager.errors.
	recordEventFunc func(eventType, event string), // typically Kubelet.recordNodeStatusEvent
	localStorageCapacityIsolation bool,
) Setter {
```

AnishShah commented:
> Rather than moving static pod registration, I sort of expected us to just track whether the existing code paths had registered static pods at least once, and gate reporting the node as ready until that happened here:

Based on the current situation, if we gate node readiness until static pods are registered, then it can increase the latency for node to become ready by 1-1.5 mins (which seems unacceptable). Hence, I feel that we need both - syncing static pods on node registration **and** blocking node readiness until static pods are registered. This will increase the latency for node readiness by  a few seconds.

I'll sync with David to see if there's a better way to do this. Converting this to draft until then.

liggitt commented:
> if we gate node readiness until static pods are registered, then it can increase the latency for node to become ready by 1-1.5 mins (which seems unacceptable)

It would be helpful to understand why. I suspect some of the things blocking the pod sync loop may be relevant and *should* block syncing static pods as well.

bobbypage commented on pkg/kubelet/kubelet.go:
```diff
@@ -2244,6 +2244,27 @@ func (kl *Kubelet) getPodsToSync() []*v1.Pod {
+		}
+	}
+
+	kl.HandlePodSyncs(staticPods)
```
hmm, I don't think here we necessarily need to wait for the full pod to sync (which can add additional latency), but rather only the mirror pod to be created. Is there any way we can check if the mirror pod is created (but avoid waiting for full sync?)

AnishShah commented:
> It would be helpful to understand why. I suspect some of the things blocking the pod sync loop may be relevant and should block syncing static pods as well.

At node startup, static pods are created and running. But mirror pod creation fails due to race condition with node registration. This is a soft failure, that is, pod is not immediately resynced. kubelet waits 1-1.5 mins to resync again ([code](https://github.com/kubernetes/kubernetes/blob/4f10b245031193c1c1679573201810ffac9191a4/pkg/kubelet/pod_workers.go#L1494)). mirror pods will be created after this resync of 1-1.5 mins. So if we just gate node readiness on mirror pod creation, it will increase the latency by 1-1.5 mins.

bobbypage commented:
> Rather than moving static pod registration, I sort of expected us to just track whether the existing code paths had registered static pods at least once, and gate reporting the node as ready until that happened here

I think that approach makes quite a bit of sense. For example, I don't think we necessarily need to wait for the pods to actually complete their full sync, only for the [mirror pods to be created](https://github.com/kubernetes/kubernetes/blob/b67d4dd389076ee1c6d12c7a41571d0b609f39ec/pkg/kubelet/kubelet.go#L1913-L1915). If we can block node to report ready until all the static pods have a corresponding mirror pod created, I think it should be sufficient.

AnishShah committed:
kubelet: try registering mirror pods as soon as node is registered.

Mirror pods for static pods may not be created immediately during node startup
because either the node is not registered or node informer is not synced.
They will be created eventually when static pods are resynced (every 1-1.5 minutes).

However, during this delay of 1-1.5 mins, kube-scheduler might overcommit resources
to the node and eventually cause kubelet to reject pods with
OutOfCPU/OutOfMemory/OutOfPods error.

To ensure kube-scheduler is aware of static pod resource usage faster,
mirror pods are created as soon as the node registers.

AnishShah commented:
/cc @tallclair

AnishShah commented:
>  As this is a bugfix, please provide unit or e2e test(s) covering it.

@bart0sh , added some unit tests. Other comments are stale as I made changes after talking to David and Tim. PTAL

> Rather than moving static pod registration, I sort of expected us to just track whether the existing code paths had registered static pods at least once, and gate reporting the node as ready until that happened here:

@liggitt @bobbypage I made changes to gate node readiness on static pod registration. But just doing this will increase the node readiness latency by 1-1.5 mins because kubelet waits another 1-1.5 mins to resync static pods. So I made changes to resync static pods immediately on node registration. Unlike previous patch, this is being done as a part of syncLoopIteration.

AnishShah commented:
We have bunch of [mirror pod e2e tests](https://github.com/kubernetes/kubernetes/blob/7b80cdb66a390f225d23cd612950144e3a39d1ae/test/e2e_node/mirror_pod_grace_period_test.go#L41) that have become disruptive tests because of my changes.

Since we are gating node readiness on static pod registration, nodes become not ready due to mirror pod e2e tests.

AnishShah commented on pkg/kubelet/kubelet.go:
```diff
@@ -401,8 +402,17 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
-			return kubeInformers.Core().V1().Nodes().Informer().HasSynced()
+			return nodeInformer.HasSynced()
+		}
+		_, err := nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```
I think it does not call addfunc for existing objects. Since it does not call for existing objects, I feel that there is no need to add an event handler. WDYT?

AnishShah commented:
/retest

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -2513,6 +2527,17 @@ func (kl *Kubelet) syncLoopIteration(ctx context.Context, configCh <-chan kubety
+		if kl.kubeClient != nil {
+			staticPods, mirrorPods := kl.podManager.GetStaticPodsAndMirrorPods()
+			if len(staticPods) != len(mirrorPods) {
+				handler.HandlePodSyncs(staticPods)
```
It's probably not good to use the main thread for creating pods at the apiserver. 

Another possible approach is to move the mirror pod creation to the beginning of the syncpod to accelerate the operations.

This would also help minimize the window for race condition for any new static pods that are added after node registration.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -1343,6 +1362,16 @@ type Kubelet struct {
+
+	// whether initial static pods registration was completed or not
+	// before node became ready.
+	staticPodsRegistered bool
```
Could you make the name more specific since static pods are not really static and can change? For example, "initialStaticPodsRegistered"

AnishShah replied:
Done.

yujuhong commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3118,23 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+		return nil
+	}
+
+	return fmt.Errorf("all static pods are not registered")
```
s/all static pods are not registered/not all static pods are registered
for clarity

AnishShah replied:
Done.

bobbypage commented on pkg/kubelet/kubelet.go:
```diff
@@ -401,8 +403,22 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
-			return kubeInformers.Core().V1().Nodes().Informer().HasSynced()
+			return nodeInformer.HasSynced()
+		}
+		handle, err := nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```
I'm curious if we need nodeInformer usage here. What specific signal are we me monitoring for registartion here?

For self registration enabled kubelets, we could check if `registerWithAPIServer` completed successfully? [1]

[1] https://github.com/kubernetes/kubernetes/blob/e9c9a27c97634c433dd9dad1118a98b7e4d519f8/pkg/kubelet/kubelet_node_status.go#L50-L50   

For non self registration kubelets, perhaps we can just check if `GetNode` ever succeded (and did not return the `initialNode`) object [2]

[2] https://github.com/kubernetes/kubernetes/blob/e9c9a27c97634c433dd9dad1118a98b7e4d519f8/pkg/kubelet/kubelet_getters.go#L279-L279

AnishShah replied:
The alternate way to do this as you mentioned would be to run a Go routine that periodically polls `GetNode`. When it succeeds, send a message over the channel to sync static pods.

Between the two choices, I felt the node informer would be faster. But maybe the Go routine approach will be cleaner?

tallclair replied:
I think I was the one that suggested this approach.

> Or is the idea that signal of node informer of node object creation will cover both of these cases?

Yes, that's the idea.

Part of the idea of using an informer with an event handler is to minimize time to node ready. Why would a polling approach be preferred?

aojea replied:
I find the informer approach simpler/easier to understand

bobbypage commented on pkg/kubelet/kubelet.go:
```diff
@@ -2513,6 +2542,19 @@ func (kl *Kubelet) syncLoopIteration(ctx context.Context, configCh <-chan kubety
 			}
 			klog.V(4).InfoS("SyncLoop (housekeeping) end", "duration", duration.Round(time.Millisecond))
 		}
+	case <-kl.nodeRegistrationCh:
```
What ensures that the mirror / static pods are handled first? Since this is part of `syncLoopIteration` won't the pods also be sent over `configCh` as well and handled there as well?

AnishShah replied:
I thought podWorker would only handle one request at a time for a particular pod. So does it matter?

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3118,23 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+		return nil
+	}
+	allStaticPods, allMirrorPods := kl.podManager.GetStaticPodsAndMirrorPods()
+	if len(allStaticPods) == len(allMirrorPods) {
```
does this guarantees all static pods are mirror or can be scenarios then len is the same but the content is different?

AnishShah replied:
yeah. If you look at `GetStaticPodsAndMirrorPods` func, we only fetch static pods and its corresponding mirror pods.

AnishShah commented:
/retest

mrunalp commented:
@haircommander @kannon92 ptal

SergeyKanzhelev commented:
/assign @mrunalp 

as @mrunalp may have way more use cases and assumptions. We need to get feedback from his experience

AnishShah commented:
Okay, I have finally found the commit (#99336) that caused this regression and leads to OutOfCPU/OutOfPods issue. 

Before #99336, During static pod sync, we blocked mirror pod creation until node was registered and node informers were in sync. mirror pods were created just ~1s after node informers were in sync. This might still be racy as node could become ready in that 1second and scheduler can overcommit pods to the node. But less likely.

```
+---------------------------+     +-------------------------+     +------------------+
|   Node Registration       |     |        SyncPod          |     |   kube-scheduler |
|     (Kubelet)             |     |       (Kubelet)         |     |                  |
+---------------------------+     +-------------------------+     +------------------+
           |                                 |  Sync static pods            |
           |                                 |                              |
           |                                 |  Block on node               |
           |                                 |  registration                |
           |                                 |<------------------------     |
           |                                 |                              |
           |  Node registered                |                              |
           |-------------------------------->|                              |
           |                                 |                              |
           |                                 |  Create mirror pod           |
           |                                 |----------------------->      |
           |                                 |                              |
           |                                 |                              |
           |                                 |         Aware of resource    |
           |                                 |         requests from        |
           |                                 |         static pod           |
           |                                 |<----------------------       |
           |                                 |                              |
           |                                 |                              |
           |                    Node ready   |                              |
           |-------------------------------------------------------------->|
           |                                 |                              |
           |                                 |             Schedule pods    |
           |                                 |             to node          |
           |<--------------------------------------------------------------|
           |                                 |                              |
```

After https://github.com/kubernetes/kubernetes/pull/99336, we stopped blocking mirror pod creation until node informers were in sync. mirror pod creation failed softly and kubelet brings up the static pod successfully. After 1-1.5 mins of sleep, kubelet syncs the static pod again and this time it is able to create mirror pods successfully. But during this sleep, the kube-scheduler ends up overcommitting to the node.

```
+---------------------------+     +------------------------+     +------------------+
|  Node Registration        |     |  SyncPod               |     |  Kube-Scheduler  |
|  (Kubelet Go routine)     |     |  (Kubelet Go routine)  |     |                  |
+---------------------------+     +------------------------+     +------------------+
           |                                |                             |
           |                                |                             |
           | Register node                  |                             |
           | with API server                |                             |
           |                                |                             |
           |                                | Sync static pods            |
           |                                |                             |
           |                                | Mirror pod creation         |
           |                                | fails softly (node          |
           |                                | informers not synced)       |
           |                                |                             |
           |                                | Static pod running          |
           |                                |                             |
           | Perform checks                 |                             |
           |                                | Sleep for 1-1.5 mins        |
           | Mark node as ready             |                             |
           | for scheduling                 |                             |
           |                                |                             |
           |                                |                        Schedule pods
           |                                |                        (node already
           |                                |                         marked ready)
           |                                |                             |
           |                                |                             |
           |                                |                    Overcommit pods
           |                                |                    to the node
           |                                |                             |
           |                                |                             |
           |                                |                             |
           |                                |                             |
           |                                |                             |
           |                                | Sync static pods again      |
           |                                |                             |
           |                                | Create mirror pod           |
           |                                |                             |
           |                                |                             |
           v                                v                             v
           ```

bart0sh commented on pkg/kubelet/kubelet.go:
```diff
@@ -1904,7 +1904,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
 		}
 		if mirrorPod == nil || deleted {
-			node, err := kl.GetNode()
+			node, err := kl.blockingGetNode()
```
```suggestion
			node, err := kl.blockingGetNode(ctx)
```

bart0sh commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -283,6 +284,24 @@ func (kl *Kubelet) GetNode() (*v1.Node, error) {
 
+// blockingGetNode ensures that the node informers are synced and returns
+// the node info for the configured node name of this kubelet.
+func (kl *Kubelet) blockingGetNode() (*v1.Node, error) {
```
```suggestion
func (kl *Kubelet) blockingGetNode(ctx context.Context) (*v1.Node, error) {
```

bart0sh commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -283,6 +284,24 @@ func (kl *Kubelet) GetNode() (*v1.Node, error) {
+// the node info for the configured node name of this kubelet.
+func (kl *Kubelet) blockingGetNode() (*v1.Node, error) {
+	if kl.kubeClient == nil {
+		return kl.initialNode(context.TODO())
```
```suggestion
		return kl.initialNode(ctx)
```

bart0sh commented on pkg/kubelet/kubelet_getters.go:
```diff
@@ -283,6 +284,24 @@ func (kl *Kubelet) GetNode() (*v1.Node, error) {
+		return kl.initialNode(context.TODO())
+	}
+	// if we have a valid kube client, we wait for initial lister to sync
+	if !kl.nodeHasSynced() {
+		for {
+			if kl.nodeHasSynced() {
+				break
+			}
+			time.Sleep(time.Second)
+		}
+	}
```
```suggestion
        for !kl.nodeHasSynced() {
		time.Sleep(time.Second)
	}
```

aojea commented:
> The latest commit reverts to the behavior before #99336 

that PR was needed because of the large startup time, 

> Without this patch: 72 seconds
> With this patch: 12 seconds
> (it restores the old fast timing / behavior)

do you mean we are going to regress on node startup now? we can not do that

liggitt commented:
> The latest commit reverts to the behavior before https://github.com/kubernetes/kubernetes/pull/99336

Agree with @aojea, we can't just revert that commit. That's a good find, but that means we need to consider a third approach that continues to run static pods immediately but shrink the time-to-report for mirror pods. I think the earlier approach of accelerating / waiting for mirror pods to report before marking the node ready was a better direction.

AnishShah commented:
> do you mean we are going to regress on node startup now? we can not do that

No, we don't regress on node startup. But we regress on static pods startup. It will be blocked until node is registered. 

> platforms where kubelet bootstrap the cluster, kube-apiserver, kube-scheduler, kube-controller manager are static pods

I see, I didn't considered this usecase. I have reverted my commit for now as it doesn't work for this.

tallclair reviewed (commented):
Sorry, I just realized I had pending comments that I never submitted...

AnishShah commented on pkg/kubelet/kubelet.go:
```diff
@@ -401,8 +403,22 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
+		if err != nil {
+			return nil, err
+		}
+		cleanupNodeEventHandler = func() {
+			if err := nodeInformer.RemoveEventHandler(handle); err != nil {
+				klog.ErrorS(err, "failed to remove event handler from node informer")
+			}
```
I wanted to do the same. But `RemoveEventHandler` requires `handler` obj which is returned by `AddEventHandler`. Thoughts?

liggitt replied:
Agreed, I don't think we should try adding/removing handlers for this. If we already have a handler registered, use that

AnishShah replied:
I couldn't find a clean way to add/remove handlers. I also didn't find any existing registered handler in the codebase. So I added a go routine that polls to ensure the node is registered and creates mirror pods after that

aojea replied:
We are using the same approach for other things so I think is not as bad, and since it only runs during the startup I think is safe

AnishShah commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3118,23 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+		return nil
+	}
+	// Check if initial static pod registration has already been completed.
+	if kl.staticPodsRegistered {
```
Done. I use `sync.Mutex` to make it safe.

AnishShah commented:
/retest

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -1886,37 +1901,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
-			}
-		}
-	}
+	kl.ensureMirrorPodExists(pod, mirrorPod)
```
```suggestion
        if kubetypes.IsStaticPod(pod) {
	     kl.ensureMirrorPodExists(pod, mirrorPod)
	 }
```

can we make this method to assume the caller only use it with static pods?

aojea replied:
it seems is called from two places, so making the check internally to be safe sounds more correct in case in the future there is some refactor that does not notice it

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3061,79 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+
+// Ensure Mirror Pod for Static Pod exists.
+func (kl *Kubelet) ensureMirrorPodExists(staticPod, mirrorPod *v1.Pod) {
+	if kubetypes.IsStaticPod(staticPod) {
```
alternative to https://github.com/kubernetes/kubernetes/pull/126870/files#r1785780245 you just remove an indentantion by returning early
```suggestion
	if !kubetypes.IsStaticPod(staticPod) {
	     return
	 }
```

aojea commented on pkg/kubelet/pod/pod_manager.go:
```diff
@@ -70,6 +70,12 @@ type Manager interface {
+	// mirror pods. It is possible that there is no mirror pod for a static pod
+	// if kubelet is running in standalone mode or is in the process of creating
+	// the mirror pod and in that case, the mirror pod is nil.
+	GetStaticPodToMirrorPodMap() map[*v1.Pod]*v1.Pod
```
This deserve an unit test

AnishShah replied:
Done.

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3061,79 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+}
+
+// Ensure Mirror Pod for Static Pod exists as soon as node is registered.
+func (kl *Kubelet) fastStaticPodsRegistration() {
```
can you please add a context here ?
```suggestion
func (kl *Kubelet) fastStaticPodsRegistration(ctx context.Context) {
```

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3061,79 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+			break
+		}
+		klog.V(5).ErrorS(err, "unable to ensure mirror pod because node is not registered yet", "node", klog.KRef("", string(kl.nodeName)))
+		time.Sleep(time.Second)
```
```suggestion
               select {
               case <-ctx.Done():
                    return
                default:
                }
		time.Sleep(time.Second)
```

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -1343,6 +1343,14 @@ type Kubelet struct {
+
+	// whether initial static pods registration was completed or not
+	// before node became ready.
+	initialStaticPodsRegistered bool
+
+	// Lock to ensure it is concurrency-safe to access
+	// initialStaticPodsRegistered field.
+	initialStaticPodsRegisteredLock sync.Mutex
```
This locks `initialStaticPodsRegistration()` entire execution, that IIUIC is not necessarily to lock the entire operation and we just may be worried about the access to the `initialStaticPodsRegistered` boolean, if that is the case, we can just use `atomic.Bool` and avoid adding more locking

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -3076,3 +3061,79 @@ func (kl *Kubelet) warnCgroupV1Usage() {
+		}
+	}
+
+	kl.initialStaticPodsRegistered = true
```
we may want to add a metric to track the time until all the static pods are registered, it is simply to add just check https://github.com/kubernetes/kubernetes/pull/121041 for reference

aojea replied:
absolutely

aojea commented:
@tallclair @AnishShah @SergeyKanzhelev let's see if we can have this early in the cycle so it gets soak time, adding too late in the release cycle is going to be tricky as this has an important impact on the node bootstrapping

rphillips commented on pkg/kubelet/kubelet.go:
```diff
@@ -1671,6 +1679,13 @@ func (kl *Kubelet) Run(updates <-chan kubetypes.PodUpdate) {
+		// created eventually when the static pods are resynced 1-1.5 mins later. However, this also affects node
+		// readiness latency as we gate node readiness on initial static pods being registered. To improve the node
+		// readiness latency, we want to ensure mirror pods exists as soon as the node is registered.
+		go kl.fastStaticPodsRegistration()
```
This function will race on execution since the static pods are not loaded yet. The static pod load, IIRC, happens in syncLoop() down below.

aojea replied:
don;t call both the same method with this change?

rphillips replied:
With the other changes that have gone in, I think this is ok with the readiness loop and cancellation context that have been added.

AnishShah commented:
/milestone v1.32

aojea commented:
LGTM

defer to node folks for final review

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+		if mirrorPod.DeletionTimestamp != nil || !kubepod.IsMirrorPodOf(mirrorPod, staticPod) {
+			// The mirror pod is semantically different from the static pod. Remove
+			// it. The mirror pod will get recreated later.
+			klog.InfoS("Trying to delete pod", "pod", klog.KObj(staticPod), "podUID", mirrorPod.ObjectMeta.UID)
```
it is strange that we are logging `staticPod` in this error message, and `mirrorPod` later. I cannot think of a problem it would cause, but inconsistency looks bad

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+	return nil
+}
+
+// Ensure Mirror Pod for Static Pod exists.
```
```suggestion
// Ensure Mirror Pod for Static Pod exists and matches the current pod definition. The function ignores any errors.
```

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+	if mirrorPod == nil || deleted {
+		node, err := kl.GetNode()
+		if err != nil {
+			klog.V(4).ErrorS(err, "No need to create a mirror pod, since failed to get node info from the cluster", "node", klog.KRef("", string(kl.nodeName)))
```
why logging is level 4? This higher level hid the issue from us. It is just a few log messages, let's make them a default level

SergeyKanzhelev replied:
same below

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for _, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil {
+			return fmt.Errorf("not all static pods are registered")
```
will it be bad to include the mirror pod name into this message?

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -1889,37 +1900,7 @@ func (kl *Kubelet) SyncPod(ctx context.Context, updateType kubetypes.SyncPodType
-			}
-		}
-	}
+	kl.ensureMirrorPodExists(pod, mirrorPod)
```
since we are ignoring all errors from `ensureMirrorPodExists`, this is just a best effort, right? We still have some room for race condition from static pods, correct?

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+			klog.V(4).InfoS("No need to create a mirror pod, since node has been removed from the cluster", "node", klog.KRef("", string(kl.nodeName)))
+		} else {
+			klog.V(4).InfoS("Creating a mirror pod for static pod", "pod", klog.KObj(staticPod))
+			if err := kl.mirrorPodClient.CreateMirrorPod(staticPod); err != nil {
```
my understanding is that this will create the pod with that API server. How fast will we get it back in the `mirrorPodByFullName`? Should we expedite resyncing?

SergeyKanzhelev replied:
Ideally we should _attempt_ to register and then mark the node as ready. I don't think we should keep waiting for mirror to be ACTUALLY created. It just opens door for all sorts of edge cases we haven't thought of

SergeyKanzhelev replied:
> > Ideally we should attempt to register and then mark the node as ready. I don't think we should keep waiting for mirror to be ACTUALLY created.
> 
> The race condition causing OutOfCPU error is because we don't wait for the mirror pod to be created before the node is ready.

My understanding is that the OutOfCPU happening because of two things - we create static pod too late, and then do not wait for it's creation. My understanding of this PR is that we are trying to fix both - create mirrors fast, and also wait for them to mark the Node as ready. I have no issues with the first, it is safe enough for me. But I am worried that the second can cause troubles if for some reasons we will enter the loop of not being able to create a mirror pod at all. Like some admission webhook misconfiguration.

So if we are doing the waiting for status I would say we need:

1. To have a max wait timeout.
2. Protection from somebody constantly recreating new static pods.
3. Ability to turn the logic completely off (perhaps a feature gate)

AnishShah replied:
> My understanding is that the OutOfCPU happening because of two things - we create static pod too late, and then do not wait for it's creation.

The issue happens because mirror pods are created too late after node has become ready. 

> My understanding of this PR is that we are trying to fix both - create mirrors fast, and also wait for them to mark the Node as ready.

Correct.

> But I am worried that the second can cause troubles if for some reasons we will enter the loop of not being able to create a mirror pod at all. Like some admission webhook misconfiguration.

If mirror pods are not created ,then it can cause more issues like more pods scheduled on the node than it can handle. We would still see OutOfCPU issue in that case?

> 1. To have a max wait timeout.

It feels weird that we consider it as a problem only for a timeout duration. What would this timeout duration be? We would still see this OutOfCPU issue after the timeout period, so we are not fixing the issue?

> 2. Protection from somebody constantly recreating new static pods.

I have added this already. In `initialStaticPodsRegistration`, we only wait until initial static pods are registered to API server. 

> 3. Ability to turn the logic completely off (perhaps a feature gate)

Since this is a bug fix, does it make sense to keep it behind a feature gate?

SergeyKanzhelev replied:
My problem here is that while fixing this issue we may introduce edge cases that will break customers in worse way than OutOfCPU. Imagine somebody with the bad static pod that somehow will fail creation. The cluster is working fine today, but with this bug fix they will end up not registering node AT ALL. This is way worse than one OutOfCPU issue, that may not be an issue for them at all.

SergeyKanzhelev replied:
This is why I offering some ways we should consider as a backup safety options

aojea replied:
> Since this is a bug fix, does it make sense to keep it behind a feature gate?

we use to rollout changes that may impact visible behavior behind a feature gate as deprecated, so if we find out a few releases later that there is a problem we have a escape hatch without having to backport and tell usesrs to upgrade to the latest minor

> If mirror pods are not created ,then it can cause more issues like more pods scheduled on the node than it can handle.

agree with @SergeyKanzhelev here, not creating mirror pods is not as a big problem as not marking the node ready,  the mirror pod is just a representation of the pod that is already running in the node and functional ... I don't think is a bad idea to have a guardrail with a timeout and fall back to the previous behavior

AnishShah replied:
To resolve this concern, I propose splitting this PR into 2 PRs:

1. This PR only contains logic to register static pods quickly once node is registered. I feel this change should be safe to merge before 1.32 deadline and help reduce OutOfCPU issue.

2. Second PR will contain the logic to block node readiness on static pod registration. We can address this post-1.32 with more protection for safer release om future.

WDYT?

SergeyKanzhelev replied:
> 1. This PR only contains logic to register static pods quickly once node is registered. I feel this change should be safe to merge before 1.32 deadline and help reduce OutOfCPU issue.

Yes, as long as the fix is reasonably contained. Definitely a good idea to move the first attempt of mirrors registration to right after the node was registered.

> 2. Second PR will contain the logic to block node readiness on static pod registration. We can address this post-1.32 with more protection for safer release om future.

Maybe if you are saying "temporarily block node readiness", it is fine. I don't think we can merge anything that will block the node readiness for a long time with the dependency on mirror pod creation being successful.

aojea replied:
I like the idea and I think that targeting 1. for this release is a good compromise ...

tallclair replied:
+1 to prioritizing fast mirror pod creation.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3051,83 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+	}
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for _, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil {
```
what if `DeletionTimestamp` if `nil`? In this case we will not even attempt to register the mirror with the API server

SergeyKanzhelev reviewed (commented):
some initial comments. Also @mrunalp , is the approach of delaying Node's `Ready` state looks OK for you?

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3052,88 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
 	return kl.containerManager.UnprepareDynamicResources(ctx, pod)
 }
+
+// staticPodRegistration ensures that all static pods are registered to the apiserver
```
```suggestion
// initialStaticPodsRegistration ensures that all static pods are registered to the apiserver
```

AnishShah commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3052,88 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+	}
+	deleted := false
+	if mirrorPod != nil {
+		if mirrorPod.DeletionTimestamp != nil || !kubepod.IsMirrorPodOf(mirrorPod, staticPod) {
```
Yes, you commented above and I added a check for `DeletionTimestamp` - https://github.com/kubernetes/kubernetes/pull/126870#discussion_r1788128543

With respect to this function, if `DeletionTimestamp` is set, then we first delete it and then create it below. So it will be cleared out.

AnishShah commented:
/assign @haircommander 
/assign @mrunalp

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,63 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+
+// Ensure Mirror Pod for Static Pod exists and matches the current pod definition.
+// The function logs and ignores any errors.
+func (kl *Kubelet) ensureMirrorPodExists(staticPod, mirrorPod *v1.Pod) {
```
This may be a better name since the function doesn't actually `ensure`, it is just trying to register and delete if needed

```suggestion
func (kl *Kubelet) tryReconcileMirrorPods(staticPod, mirrorPod *v1.Pod) {
```

AnishShah replied:
Done.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,63 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+			if err != nil {
+				klog.ErrorS(err, "Failed deleting mirror pod", "pod", klog.KObj(mirrorPod))
+			} else if deleted {
+				klog.InfoS("Deleted mirror pod", "pod", klog.KObj(mirrorPod))
```
```suggestion
				klog.InfoS("Deleted mirror pod as it didn't match the static Pod", "pod", klog.KObj(mirrorPod))
```

AnishShah replied:
Done.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,63 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+			}
+		}
+	}
+	if mirrorPod == nil || deleted {
```
You cannot use the value of `deleted` here. If `err` was returned from `kl.mirrorPodClient.DeleteMirrorPod`, we cannot rely on the value of `deleted`

AnishShah replied:
Added a `deleteErr == nil` check.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,63 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+		} else {
+			klog.InfoS("Creating a mirror pod for static pod", "pod", klog.KObj(staticPod))
+			if err := kl.mirrorPodClient.CreateMirrorPod(staticPod); err != nil {
+				klog.ErrorS(err, "Failed creating a mirror pod for", "pod", klog.KObj(staticPod))
```
```suggestion
				klog.ErrorS(err, "Failed creating a mirror pod", "pod", klog.KObj(staticPod))
```

AnishShah replied:
Done.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,63 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+		case <-ctx.Done():
+			return
+		case <-t.C:
+		}
```
let's add an `klog.InfoS` here on every tick. So it is easier to debug if something is not going the way we expect it to go.

SergeyKanzhelev replied:
in success case - we only see a couple of ticks. In case when node is not registered, kubelet will not live long enough to produce too many logs

AnishShah replied:
Now that I'm using `PollUntilContextCancel`, this is not needed anymore?

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,63 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+func (kl *Kubelet) fastStaticPodsRegistration(ctx context.Context) {
+	t := time.NewTimer(time.Second)
+	defer t.Stop()
+	for {
+		_, err := kl.GetNode()
+		if err == nil {
+			break
+		}
+		klog.V(5).ErrorS(err, "Unable to ensure mirror pod because node is not registered yet", "node", klog.KRef("", string(kl.nodeName)))
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+		}
```
You can use a helper method `PollUntilContextCancel` instead of dealing with channels. Less code, less potential for errors

AnishShah replied:
Thanks. Done.

AnishShah commented:
/retest

aojea commented:
/lgtm

previous error looks unrelated
```
{Failed;Failed;  === RUN   TestAuthorizerMetrics/canceled_request
    metrics_test.go:140: incorrect status code recorded for RecordRequestTotal method, expected = <error>, got 200
    metrics_test.go:144: incorrect status code recorded for RecordRequestLatency method, expected = <error>, got 200
    metrics_test.go:148: expected evaluationsResult "canceled", got "success"
    metrics_test.go:151: expected durationResult "canceled", got "success"
--- FAIL: TestAuthorizerMetrics/canceled_request (0.08s)
```

haircommander commented:
LGTM I like this approach

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -3070,3 +3047,61 @@ func (kl *Kubelet) PrepareDynamicResources(ctx context.Context, pod *v1.Pod) err
+			}
+		}
+	}
+	if mirrorPod == nil || (deleted && deleteErr == nil) {
```
why do you need a `deleteErr` here?

AnishShah replied:
You mentioned in https://github.com/kubernetes/kubernetes/pull/126870#discussion_r1819612251 that we cannot rely on `deleted` value if `deletedErr` is not nil.

SergeyKanzhelev replied:
In a past iteration you didn't have `if ok, err := kl.mirrorPodClient.DeleteMirrorPod(podFullName, &mirrorPod.ObjectMeta.UID); err != nil {` and instead simply initialized the `deleted` something like this: `deleted, err = kl.mirrorPodClient.DeleteMirrorPod(podFullName, &mirrorPod.ObjectMeta.UID)`

Now you only initialize deleted when error is not nil. So it's fine to rely on it's value

AnishShah replied:
Thanks. Done.

SergeyKanzhelev commented on pkg/kubelet/pod/pod_manager_test.go:
```diff
@@ -294,3 +294,15 @@ func TestRemovePods(t *testing.T) {
+		t.Fatalf("GetStaticPodToMirrorPodMap(): got %d static pods, wanted 1 static pod", len(m))
+	}
+	if gotMirrorPod, ok := m[staticPod]; !ok || gotMirrorPod.UID != mirrorPod.UID {
+		t.Fatalf("GetStaticPodToMirrorPodMap() did not return the correct mirror pod UI %s, wanted mirror pod UID %s", gotMirrorPod.Name, mirrorPod.Name)
```
```suggestion
		t.Fatalf("GetStaticPodToMirrorPodMap() did not return the correct mirror pod UI %s, wanted mirror pod UID %s", gotMirrorPod.UID, mirrorPod.UID)
```

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

AnishShah commented:
/retest

aojea commented:
awesome

SergeyKanzhelev commented:
> awesome

cannot agree more
